src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+            {
+                double dimEntropy = GetEntropy(pointDis[key], anomalyDis[key]);
+                entropy += dimEntropy * pointDis[key] / total;
+            }
+
+            return entropy;
+        }
+
+        private double GetDimensionInstrinsicValue(Dictionary<string, int> pointDis)
+        {
+            int total = GetTotalNumber(pointDis);
+            double instrinsicValue = 0;
+
+            foreach (string key in pointDis.Keys)
+            {
+                instrinsicValue -= Log2((double)pointDis[key] / total) * (double)pointDis[key] / total;
+            }
+
+            return instrinsicValue;
+        }
+
+        private int GetTotalNumber(Dictionary<string, int> distribution)
+        {
+            int total = 0;
+            foreach (int num in distribution.Values)
+            {
+                total += num;
+            }
+            return total;
+        }
+
+        private void UpdateDistribution(Dictionary<string, int> distribution, List<Point> points, string dimKey)
+        {
+            foreach (Point point in points)
+            {
+                string dimVal = (string)point.Dimension[dimKey];
+                if (!distribution.ContainsKey(dimVal))
+                {
+                    distribution.Add(dimVal, 0);
+                }
+                distribution[dimVal] = distribution[dimVal] + 1;
+            }
+        }
+
+        private double Log2(double val)
+        {
+            if (Double.IsNaN(val))
+            {
+                return 0;
+            }
+
+            return Math.Log(val) / Math.Log(2);
+        }
+
+        private static bool ContainsAll(Dictionary<string, Object> bigDic, Dictionary<string, Object> smallDic)
+        {
+            foreach (var item in smallDic)
+            {
+                if (!bigDic.ContainsKey(item.Key))
+                {
+                    return false;
+                }
+
+                if (bigDic.ContainsKey(item.Key) && !bigDic[item.Key].Equals(smallDic[item.Key]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool IsAggregationDimension(Object val, Object aggSymbol)
+        {
+            return val.Equals(aggSymbol);
+        }
+    }
+
+    public class DimensionInfo
+    {
+        public List<string> DetailDims { get; set; }
+        public List<string> AggDims { get; set; }
+
+        public static DimensionInfo CreateDefaultInstance()
+        {
+            DimensionInfo instance = new DimensionInfo();
+            instance.DetailDims = new List<string>();
+            instance.AggDims = new List<string>();
+            return instance;
+        }
+    }
+
+    public class PointTree
+    {
+        public Point ParentNode;
+        public Dictionary<string, List<Point>> ChildrenNodes;
+        public List<Point> Leaves;
+
+        public static PointTree CreateDefaultInstance()
+        {
```

Comment:
question: Same question as above and also below for BestDimension. Can you use a regular constructor?

Reply:
> question: Same question as above and also below for BestDimension. Can you use a regular constructor?

updated